File: C++/String/smith_waterman_algo.cpp
#include <iostream>
#include <vector>

int max(int a, int b, int c, int d) {
    int max = a;
    if (b > max) {
        max = b;
    }
    if (c > max) {
        max = c;
    }
    if (d > max) {
        max = d;
    }
    return max;
}

void smith_waterman(const std::string& seq1, const std::string& seq2, int match_score, int mismatch_score, int gap_penalty) {
    int rows = seq1.length() + 1;
    int cols = seq2.length() + 1;

    std::vector<std::vector<int>> matrix(rows, std::vector<int>(cols, 0));

    // Fill the scoring matrix
    int max_score = 0;
    int max_i = 0;
    int max_j = 0;
    for (int i = 1; i < rows; i++) {
        for (int j = 1; j < cols; j++) {
            int match = (seq1[i - 1] == seq2[j - 1]) ? matrix[i - 1][j - 1] + match_score : matrix[i - 1][j - 1] + mismatch_score;
            int delete_score = matrix[i - 1][j] + gap_penalty;
            int insert_score = matrix[i][j - 1] + gap_penalty;
            int score = max(match, delete_score, insert_score, 0);

            if (score > max_score) {
                max_score = score;
                max_i = i;
                max_j = j;
            }

            matrix[i][j] = score;
        }
    }

    // Traceback to find the alignment
    std::string aligned_seq1, aligned_seq2;
    int i = max_i;
    int j = max_j;
    while (matrix[i][j] != 0) {
        if (seq1[i - 1] == seq2[j - 1]) {
            aligned_seq1 = seq1[i - 1] + aligned_seq1;
            aligned_seq2 = seq2[j - 1] + aligned_seq2;
            i--;
            j--;
        } else if (matrix[i][j] == matrix[i - 1][j - 1] + mismatch_score) {
            aligned_seq1 = seq1[i - 1] + aligned_seq1;
            aligned_seq2 = seq2[j - 1] + aligned_seq2;
            i--;
            j--;
        } else if (matrix[i][j] == matrix[i - 1][j] + gap_penalty) {
            aligned_seq1 = seq1[i - 1] + aligned_seq1;
            aligned_seq2 = '-' + aligned_seq2;
            i--;
        } else {
            aligned_seq1 = '-' + aligned_seq1;
            aligned_seq2 = seq2[j - 1] + aligned_seq2;
            j--;
        }
    }

    std::cout << "Alignment 1: " << aligned_seq1 << std::endl;
    std::cout << "Alignment 2: " << aligned_seq2 << std::endl;
    std::cout << "Score: " << max_score << std::endl;
}

int main() {
    std::string sequence1 = "ACGTCG";
    std::string sequence2 = "AGTCAG";
    int match_score = 2;
    int mismatch_score = -1;
    int gap_penalty = -1;

    smith_waterman(sequence1, sequence2, match_score, mismatch_score, gap_penalty);

    return 0;
}

File: Java/Dynamic-Programming/Smith_waterman_algo.java
public class smith_waterman_algo {
    public static void main(String[] args) {
        String sequence1 = "ACGTCG";
        String sequence2 = "AGTCAG";
        int matchScore = 2;
        int mismatchScore = -1;
        int gapPenalty = -1;

        smithWaterman(sequence1, sequence2, matchScore, mismatchScore, gapPenalty);
    }

    public static void smithWaterman(String seq1, String seq2, int matchScore, int mismatchScore, int gapPenalty) {
        int rows = seq1.length() + 1;
        int cols = seq2.length() + 1;

        int[][] matrix = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            matrix[i][0] = 0;
        }
        for (int j = 0; j < cols; j++) {
            matrix[0][j] = 0;
        }

        int maxScore = 0;
        int maxI = 0;
        int maxJ = 0;
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                int match = (seq1.charAt(i - 1) == seq2.charAt(j - 1)) ? matrix[i - 1][j - 1] + matchScore : matrix[i - 1][j - 1] + mismatchScore;
                int delete = matrix[i - 1][j] + gapPenalty;
                int insert = matrix[i][j - 1] + gapPenalty;
                int score = Math.max(match, Math.max(delete, Math.max(insert, 0)));

                if (score > maxScore) {
                    maxScore = score;
                    maxI = i;
                    maxJ = j;
                }

                matrix[i][j] = score;
            }
        }

        StringBuilder alignedSeq1 = new StringBuilder();
        StringBuilder alignedSeq2 = new StringBuilder();
        int i = maxI;
        int j = maxJ;
        while (matrix[i][j] != 0) {
            if (seq1.charAt(i - 1) == seq2.charAt(j - 1)) {
                alignedSeq1.insert(0, seq1.charAt(i - 1));
                alignedSeq2.insert(0, seq2.charAt(j - 1));
                i--;
                j--;
            } else if (matrix[i][j] == matrix[i - 1][j - 1] + mismatchScore) {
                alignedSeq1.insert(0, seq1.charAt(i - 1));
                alignedSeq2.insert(0, seq2.charAt(j - 1));
                i--;
                j--;
            } else if (matrix[i][j] == matrix[i - 1][j] + gapPenalty) {
                alignedSeq1.insert(0, seq1.charAt(i - 1));
                alignedSeq2.insert(0, '-');
                i--;
            } else {
                alignedSeq1.insert(0, '-');
                alignedSeq2.insert(0, seq2.charAt(j - 1));
                j--;
            }
        }

        System.out.println("Alignment 1: " + alignedSeq1.toString());
        System.out.println("Alignment 2: " + alignedSeq2.toString());
        System.out.println("Score: " + maxScore);
    }
}
